src/HDF5Msg.chpl

```diff
@@ -0,0 +1,2193 @@
+            h5Logger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                    "Filedomains: %t".format(filedomains));
+            h5Logger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                    "skips: %t".format(skips));
+
+            coforall loc in A.targetLocales() do on loc {
+                // Create local copies of args
+                var locFiles = filenames;
+                var locFiledoms = filedomains;
+                /* On this locale, find all files containing data that belongs in
+                 this locale's chunk of A */
+                for (filedom, filename) in zip(locFiledoms, locFiles) {
+                    var isopen = false;
+                    var file_id: C_HDF5.hid_t;
+                    var dataset: C_HDF5.hid_t;
+
+                    if (skips.contains(filename)) {
+                        h5Logger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                                "File %s does not contain data for this dataset, skipping".format(filename));
+                    } else {
+                        // Look for overlap between A's local subdomains and this file
+                        for locdom in A.localSubdomains() {
+                            const intersection = domain_intersection(locdom, filedom);
+                            if intersection.size > 0 {
+                                // Only open the file once, even if it intersects with many local subdomains
+                                if !isopen {
+                                    file_id = C_HDF5.H5Fopen(filename.c_str(), C_HDF5.H5F_ACC_RDONLY, 
+                                                                                            C_HDF5.H5P_DEFAULT);  
+                                    var locDsetName = try! getReadDsetName(file_id,dsetName);
+                                    try! dataset = C_HDF5.H5Dopen(file_id, locDsetName.c_str(), C_HDF5.H5P_DEFAULT);
+                                    isopen = true;
+                                }
+                                // do A[intersection] = file[intersection - offset]
+                                var dataspace = C_HDF5.H5Dget_space(dataset);
+                                var dsetOffset = [(intersection.low - filedom.low): C_HDF5.hsize_t];
+                                var dsetStride = [intersection.stride: C_HDF5.hsize_t];
+                                var dsetCount = [intersection.size: C_HDF5.hsize_t];
+                                C_HDF5.H5Sselect_hyperslab(dataspace, C_HDF5.H5S_SELECT_SET, c_ptrTo(dsetOffset), 
+                                                                c_ptrTo(dsetStride), c_ptrTo(dsetCount), nil);
+                                var memOffset = [0: C_HDF5.hsize_t];
+                                var memStride = [1: C_HDF5.hsize_t];
+                                var memCount = [intersection.size: C_HDF5.hsize_t];
```

Comment:
@glitch: While coming across this code today, I wanted to point out a potential misunderstanding and opportunity to clean up / optimize / simplify this code (where it looks like this code is from @mhmerrill and the dawn of time, so I'm only tagging you for your info and as the last one to touch it rather than because you wrote it to begin with :) ).

From what I can tell, these declarations are creating 1-element arrays and then passing a pointer to the first/only element in the array into the HDF5 routines.  My guess is that this is because Mike believed that `c_ptrTo()` was only supported on arrays, but it's actually supported on any Chapel expression with storage.   So where one of these uses:

```chapel
var arr = [elt: t];
...c_ptrTo(arr)...
```

it could simply use:

```chapel
var val = elt: t;
...c_ptrTo(val)...
```

The savings would be the reduction in time to create an array and its governing domain for a single element ("Why play an ace when a two will do?").  

Specifically, `c_ptrTo()` should be able to be applied to any Chapel expression that has storage where arrays are a special case (in that it points to the array's initial element rather than to the array descriptor itself).

Since this is I/O code, my guess is that these O(1) overheads are washed away by the O(n) I/O cost and won't show any significant benefit from changing.  But it still seemed worth pointing out, if not for code cleanup purposes then to stem the proliferation of this idiom when it's not needed (e.g., if you're modifying or reviewing code going forward that uses it).

I also saw a variation on this pattern elsewhere in the code where the array was declared in a more verbose way:

```chapel
var A: [0..#1] t;
A[0] = val;
...c_ptrTo(A)...
```

which looked similarly like overkill.

It _could_ be that I'm missing some other subtle reason that these want to be arrays, but @bmcdonald3 took a look as well and came to similar conclusions (and I believe tried replacing some of them with scalars and things seemed to continue to work fine).